### PR TITLE
bpo-45354: Skip obsolete device name tests on Windows 11

### PR DESCRIPTION
**Pull request by jkloth**
https://bugs.python.org/issue45354

### CODE REVIEW

File: Lib/test/test_winconsoleio.py [hunk 1]
```diff
@@ -92,9 +92,11 @@ def test_open_name(self):
         f.close()
         f.close()
 
-        f = open('C:/con', 'rb', buffering=0)
-        self.assertIsInstance(f, ConIO)
-        f.close()
+        # Windows 11 changed MS-DOS device name handling
+        if sys.getwindowsversion()[:3] < (10, 0, 22000):
+            f = open('C:/con', 'rb', buffering=0)
```
Comment (eryksun) [inline Lib/test/test_winconsoleio.py:97]: What does `nt._getfullpathname('C:/con').lower()` return in Windows 11? The test should be skipped if it's not `r"\\.\con"`. Similarly, if `nt._getfullpathname(conout_path).lower()` isn't `r"\\.\conout$"`, assume that `conout_path` is a regular file.

Comment (jkloth) [reply Lib/test/test_winconsoleio.py:97]: Both calls return a normalized pathname, not the device object name.

That said, as the tests in question are originally from you, shouldn't the tests then be written in terms of _getfullpathname() ?

Comment (jkloth) [reply Lib/test/test_winconsoleio.py:97]: Meaning, instead of using open(), just check the results of _getfullpathname()

Comment (eryksun) [reply Lib/test/test_winconsoleio.py:97]: It's checking `_PyIO_get_console_type()` in "Modules/_io/winconsoleio.c", which may need to be modified. Please also check `nt._getfullpathname()` and `nt._getfinalpathname()` (may fail) on the unqualified base names "con", "conin$", and "conout$". These names resolve to the console in previous Windows versions, so `_PyIO_get_console_type()` skips calling `GetFullPathNameW()`. This optimization may need to be removed if the API no longer special cases the base names.

Comment (eryksun) [reply Lib/test/test_winconsoleio.py:97]: I verified the new behavior in Windows 11. If the path is qualified (i.e. contains slashes), only the "nul" device name appears to be reserved (e.g. "./nul" -> "\\\\.\\nul"). The other DOS device names aren't reserved in a qualified path (e.g. "./com1" is just a regular file in the current working directory). DOS device names are still reserved in an unqualified relative path, but extensions are no longer ignored (e.g. "con.txt" is just a regular file).

This means that `_PyIO_get_console_type()` doesn't need to be modified. It only special cases an exact case-insensitive match of "con", "conin$", and "conout$", which are still reserved paths.

Comment (jkloth) [reply Lib/test/test_winconsoleio.py:97]: Sooooo..... does this mean that these changes are suitable?  Does it just need an OK from @zooba or @zware?  I'm only slightly impatient here as this is blocking the new buildbot from being actually useful.

File: Lib/test/test_winconsoleio.py [hunk 2]
```diff
@@ -92,9 +92,11 @@ def test_open_name(self):
...
+            self.assertIsInstance(f, ConIO)
+            f.close()
```
Comment (vstinner) [inline Lib/test/test_winconsoleio.py:99]: Can you please modify the test to use "with f: ..."?

File: Lib/test/test_winconsoleio.py [hunk 3]
```diff
@@ -92,9 +92,11 @@ def test_open_name(self):
         f.close()
         f.close()
 
-        f = open('C:/con', 'rb', buffering=0)
-        self.assertIsInstance(f, ConIO)
-        f.close()
+        # Windows 11 changed MS-DOS device name handling
```
Comment (vstinner) [inline Lib/test/test_winconsoleio.py]: Can you add a reference to the bpo here and in the other modified test?

```suggestion
        # bpo-45354: Windows 11 changed MS-DOS device name handling
```

File: Lib/test/test_winconsoleio.py (hunk 2 above)
Comment (jkloth) [reply Lib/test/test_winconsoleio.py:99]: If you notice, the rest of the test cases do not do that.  Changing to context managers should probably be done as another issue.  The only change here is moving the existing test into the if-block.

File: Lib/test/test_winconsoleio.py (hunk 3 above)
Comment (jkloth) [reply Lib/test/test_winconsoleio.py]: Done in latest commit

File: Lib/test/test_winconsoleio.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/test_winconsoleio.py:99]: I see. Leave it as it is.

File: Lib/test/test_winconsoleio.py [hunk 4]
```diff
@@ -92,9 +92,11 @@ def test_open_name(self):
         f.close()
         f.close()
 
-        f = open('C:/con', 'rb', buffering=0)
-        self.assertIsInstance(f, ConIO)
-        f.close()
+        # bpo-45354: Windows 11 changed MS-DOS device name handling
+        if sys.getwindowsversion()[:3] < (10, 0, 22000):
```
Comment (zooba) [inline Lib/test/test_winconsoleio.py:96]: I suspect we'll want this to become a constant in the test suite soon enough, but doesn't have to be in this PR. Let's just keep an eye out for the next time it gets copied somewhere and then insist it gets centralised.

**Comment by zooba** [comment]
Ignoring the x86 test failure - it's clearly unrelated to this change.

**Pull request merged**